beacon-chain/blockchain/service.go

```diff
@@ -244,36 +183,58 @@ func (c *ChainService) blockProcessing(processedBlock chan<- *pb.BeaconBlock) {
 
 			currentSlot := beaconState.Slot
 			if currentSlot+1 < block.Slot {
-				c.unProcessedBlocks[block.Slot] = block
+				log.Debugf("Received block with slot higher than current slot + 1: %d", block.Slot)
 				continue
 			}
 
 			if currentSlot+1 == block.Slot {
-				if err := c.receiveBlock(block, beaconState); err != nil {
-					log.Error(err)
-					processedBlock <- nil
+				computedState, err := c.ReceiveBlock(block, beaconState)
+				if err != nil {
+					log.Errorf("Could not process received block: %v", err)
+					continue
+				}
+				if err := c.ApplyForkChoiceRule(block, computedState); err != nil {
+					log.Errorf("Could not update chain head: %v", err)
 					continue
 				}
-				// Push the block to trigger the fork choice rule.
```

Comment:
This no longer needs to be sent through a channel, as there is no concurrency needed. The flow of information will always be process block and immediately after, apply fork choice.